tt_eager/tt_dnn/op_library/moreh_sum/moreh_sum_op.cpp

```diff
@@ -81,69 +148,60 @@ void MorehSum::validate_with_output_tensors(
-            input_shape_wo_padding[this->dim] = 1;
-        } else {
-            input_shape[this->dim] = 1;
-            input_shape_wo_padding[this->dim] = 1;
-        }
-
-        log_debug(LogOp, "{}:{} input_shape {}", __func__, __LINE__, input_shape);
-        log_debug(LogOp, "{}:{} output_shape {}", __func__, __LINE__, output_shape);
-        log_debug(LogOp, "{}:{} input_shape_wo_padding {}", __func__, __LINE__, input_shape_wo_padding);
-        log_debug(LogOp, "{}:{} output_shape_wo_padding {}", __func__, __LINE__, output_shape_wo_padding);
-        for (int i = 0; i < input_shape.rank(); ++i) {
-            TT_FATAL(input_shape[i] == output_shape[i]);
-            TT_FATAL(input_shape_wo_padding[i] == output_shape_wo_padding[i]);
-        }
+        validate_output_tensor_with_keepdim(input, output.value(), this->dim, this->keepdim);
     }
-#endif
 }
 
 std::vector<Shape> MorehSum::compute_output_shapes(const std::vector<Tensor>& input_tensors) const {
     const auto& input = input_tensors.at(0);
     const auto& input_shape = input.get_legacy_shape();
     const auto input_rank = input_shape.rank();
+    const bool is_tile_dim = (this->dim == input_rank - 1 || this->dim == input_rank - 2);
+    log_debug(LogOp, "{}:{} dim {}, keepdim {}", __func__, __LINE__, this->dim, this->keepdim);
 
-    // keepdim=true
-    auto output_shape = input_shape;
-    auto padding = output_shape.padding();
+    Shape output_shape = input_shape;
+    if (this->keepdim) {
+        auto shape = input_shape;
+        auto padding = shape.padding();
+
+        if (is_tile_dim) {
+            // e.g. (2, 64, 64) with dim 1 to be (2, 1[32], 64)
+            shape[this->dim] = TILE_HEIGHT;
+            padding[this->dim] = Padding::PadDimension{0, 31};
+        } else {
+            // e.g. (2, 64, 64) with dim 0 to be (1, 64, 64)
+            shape[this->dim] = 1;
+        }
 
-    // last 2-dim
-    if (this->dim == input_rank - 1 || this->dim == input_rank - 2) {
-        output_shape[this->dim] = TILE_HEIGHT;
-        padding[this->dim] = Padding::PadDimension{0, 31};
+        output_shape = Shape(shape, padding);
     } else {
-        output_shape[this->dim] = 1;
+        std::vector<uint32_t> shape;
+        std::vector<Padding::PadDimension> pad_dimensions;
+        const std::size_t output_rank = (is_tile_dim) ? (input_rank) : (input_rank - 1);
+        auto input_padding = input_shape.padding();
+
+        // e.g. (2, 64, 64) with dim 1 to be (2, 1[32], 64)
+        // e.g. (2, 64, 64) with dim 0 to be (64, 64)
+        for (int i = 0; i < input_rank; ++i) {
+            bool is_reduced_dim = (i == this->dim);
+            if (is_reduced_dim && !is_tile_dim)
+                continue;
+
+            shape.push_back((is_reduced_dim && is_tile_dim) ? (TILE_HEIGHT) : (input_shape[i]));
+            pad_dimensions.push_back(
+                (is_reduced_dim && is_tile_dim) ? (Padding::PadDimension{0, 31}) : (input_padding[i]));
```

Comment:
What about making error when ```keep_dim=True && is_reduced_dim```?
Automatically turning off ```keep_dim``` could be confusing to users.